spec/syntax.md

```diff
@@ -0,0 +1,598 @@
+        _ [You bought {$countInt} {$color: adjective accord=$itemAcc} {$itemAcc}.]
+
+### Complex Messages
+
+A complex message with 2 selectors and 3 local variable definitions:
+
+    {$host: gender}
+    {$guestOther: number}
+
+    $hostName = {$host: person firstName=long}
+    $guestName = {$guest: person firstName=long}
+    $guestsOther = {$guestCount: number offset=1}
+
+        female 0 [{$hostName} does not give a party.]
+        female 1 [{$hostName} invites {$guestName} to her party.]
+        female 2 [{$hostName} invites {$guestName} and one other person to her party.]
+        female _ [{$hostName} invites {$guestName} and {$guestsOther} other people to her party.]
+
+        male 0 [{$hostName} does not give a party.]
+        male 1 [{$hostName} invites {$guestName} to his party.]
+        male 2 [{$hostName} invites {$guestName} and one other person to his party.]
+        male _ [{$hostName} invites {$guestName} and {$guestsOther} other people to his party.]
+
+        _ 0 [{$hostName} does not give a party.]
+        _ 1 [{$hostName} invites {$guestName} to their party.]
+        _ 2 [{$hostName} invites {$guestName} and one other person to their party.]
+        _ _ [{$hostName} invites {$guestName} and {$guestsOther} other people to their party.]
+
+## Comparison with ICU MessageFormat 1.0
+
+MessageFormat 2.0 improves upon the ICU MessageFormat 1.0 syntax through the following changes:
+
+1. In MessageFormat 2.0,
+   variants can only be defined at the top level of the message,
+   thus precluding any possible nestedness of expressions.
+
+   ICU MessageFormat 1.0:
+
+   ```
+   {foo, func,
+       foo1 {Value 1},
+       foo2 {
+           {bar, func,
+               bar1 {Value 2a}
+               bar2 {Value 2b}}}}
+   ```
+
+   MessageFormat 2.0:
+
+   ```
+   {$foo: func} {$bar: func}
+       foo1 [Value 1]
+       foo2 bar1 [Value 2a]
+       foo2 bar2 [Value 2b]
+   ```
+
+1. MessageFormat 2.0 differentiates between
+   the syntax used to introduce expressions (`{...}`) and
+   the syntax used to defined translatable content (`[...]`).
+
+1. MessageFormat 2.0 uses the dollar sign (`$`) as the sigil for variable references,
+   the colon (`:`) as the function call syntax,
+   and only allows named options to functions.
+   The purpose of this change is to help disambiguate between
+   the different parts of a placeholder (variable references, function names, literals etc.).
+
+   ICU MessageFormat 1.0:
+
+   ```
+   {when, date, short}
+   ```
+
+   MessageFormat 2.0:
+
+   ```
+   {$when: date style=short}
+   ```
+
+1. MessageFormat 2.0 doesn't provide the `#` shorthand inside variants.
+   Instead it allows local variables to be defined in the preamble,
+   which can then be referred to inside patterns similar to other variables.
+
+1. MessageFormat 2.0 doesn't require commas (`,`) inside placeholders.
+
+## Productions
+
+The specification defines the following grammar productions.
```

Comment:
Note / nitpick: there are currently quite a few things that are parsed but I would not consider valid.

Not actionable or blocker for this PR.
Unless you want to add a note to say "satisfying all rules of the final grammar" (adding "final")

For example `[Foo] one [Bar]` parses (as `[ TOKEN("["), Text("Foo"), TOKEN("]") ]`)
But that is not the only example.

Reply:
That might be an artifact of the generated parser, as it does not consider trailing content to be an error.

Reply:
I have already marked this as is a nitpick, and suggested a change.
If you think it improves things, fine, implement it.
If not, fine too, close it.